Index: xapian-core/net/tcpclient.cc
===================================================================
--- xapian-core/net/tcpclient.cc	(revision 14810)
+++ xapian-core/net/tcpclient.cc	(working copy)
@@ -109,7 +109,7 @@
 #endif
 	    int saved_errno = socket_errno(); // note down in case close hits an error
 	    close_fd_or_socket(socketfd);
-	    throw Xapian::NetworkError("Couldn't connect", saved_errno);
+	    throw Xapian::NetworkError("Couldn't connect (1)", saved_errno);
 	}
 
 	// wait for input to be available.
@@ -121,11 +121,19 @@
 	tv.tv_sec = msecs_timeout_connect / 1000;
 	tv.tv_usec = msecs_timeout_connect % 1000 * 1000;
 
-	retval = select(socketfd + 1, 0, &fdset, &fdset, &tv);
+	do {
+	    retval = select(socketfd + 1, 0, &fdset, &fdset, &tv);
+	} while (retval < 0 && errno == EINTR); // FIXME: for non-Linux we should adjust tv.
 
-	if (retval == 0) {
+	if (retval < 0) {
+	    int saved_errno = errno;
+	    close_fd_or_socket(socketfd);
+	    throw Xapian::NetworkError("Couldn't connect (2)", saved_errno);
+	}
+
+	if (retval <= 0) {
 	    close_fd_or_socket(socketfd);
-	    throw Xapian::NetworkTimeoutError("Couldn't connect", ETIMEDOUT);
+	    throw Xapian::NetworkTimeoutError("Timed out waiting to connect", ETIMEDOUT);
 	}
 
 	int err = 0;
@@ -143,7 +151,7 @@
 	}
 	if (err) {
 	    close_fd_or_socket(socketfd);
-	    throw Xapian::NetworkError("Couldn't connect", err);
+	    throw Xapian::NetworkError("Couldn't connect (3)", err);
 	}
     }
 
