--- xapian-core-1.2.0.orig/configure.ac
+++ xapian-core-1.2.0/configure.ac
@@ -42,7 +42,7 @@
 dnl test x"$program_suffix" != xNONE || program_suffix=-1.1
 
 dnl Where xapian.h, etc go.
-incdir=$includedir/xapian
+incdir=$includedir
 AC_SUBST(incdir)
 
 dnl Check the build directory doesn't contain a space, so we die early with
--- xapian-core-1.2.0.orig/configure
+++ xapian-core-1.2.0/configure
@@ -2801,7 +2801,7 @@
 
 
 
-incdir=$includedir/xapian
+incdir=$includedir
 
 
 case `pwd` in
